postmortems/2023-03-20_azure_dev_outage.md

```diff
@@ -0,0 +1,54 @@
+* -27 13:08 Martin reports on Teams site is back up and starts validating
+* -27 Martin runs database migration to allow app to come back up
+* -27 13:15 Martin reports on Teams incident resolved
+* -27 13:25 Martin emails G to confirm resolution and indicate a new ticket is not after all required
+
+
+## Why
+
+Original deployment was intended as a proof of concept, it's a 'dev' resource group which had a planned migration to another environment. Since then the plan for productionising has been revised, and proper domain was given to the site to enable early testing with some groups within the client organisation, which give the appearance of a more mature solution. This means both monitoring and processes around management of the environment are less robust than for the primary deployment in our own setup.
+
+The original method for obtaining the secret was not well understood by us or the supplier, we originally walked one of their support agents through the process in the Azure portal on a video call, and were then emailed the result. That secret had lifetime of one year (likely a default) and was included in one of the emails, but not added to a calendar or otherwise recorded.
+
+There is no process in the client organisation for updating secrets. Neither Martin or JJ had ability to file the ticket under the options available, and had to go through helpdesk, and the ticket had to be filed as a "Non Standard Request". There were then multiple layers within the client organisation to navigate to get the supplier to look at this outside their 10 day window for work items not otherwise categorised. With chasing, it took three days for a response.
+
+## Wonderings
+
+There was no monitoring to detect the outage, it was only spotted when trying to access the service during working hours. Again due to having been created as a temporary deployment, and also restrictions from the client around our standard tooling. The web service exposes Prometheus metrics but we don't have a service principal would let us scrape, dashboard, and alert on them. We need to decide on an approach here as part of productionisation that would catch auth specific outages. Whitebox monitoring would not be sufficient.
+
+The component serving the 500 response (the easyauth 'sidecar' component) did not return a response with any useful details as to why it was failing, nor log anywhere that could be seen from within the azure portal or by sshing into an app service. This made discovering why authentication was failing difficult. At some point, we may have to switch to handling oauth within our web server for better observability and flexibility on handling.
+
+Our approach for infrequent non-automatable processes, such as awkward cert updates, is to have a documented playbook and a calendar invite reminder. Any deployment secrets required for Azure will need to use this framework.
```

Comment:
There's something here about needing to take into account a minimum 3 day response time for an expedited support request, probably much longer for a normal one.

Reply:
"calendar reminder sufficiently in advance"?

Reply:
Although how do we determine what that means on a case by case basis? We need a rule where if we're reliant on an external provider to give us critical pieces of infrastructure, we need to be aware of their typical response time and have that written down somewhere.